README.md
# date-format-parse

<a href="https://travis-ci.org/mengxiong10/date-format-parse">
  <img src="https://travis-ci.org/mengxiong10/date-format-parse.svg?branch=master" alt="build:passed">
</a>
<a href="https://coveralls.io/github/mengxiong10/date-format-parse">
  <img src="https://coveralls.io/repos/github/mengxiong10/date-format-parse/badge.svg?branch=master&service=github" alt="Badge">
</a>
<a href="https://www.npmjs.com/package/date-format-parse">
  <img src="https://img.shields.io/npm/v/date-format-parse.svg" alt="npm">
</a>
<a href="LICENSE">
  <img src="https://img.shields.io/badge/License-MIT-yellow.svg" alt="MIT">
</a>

Lightweight date format and parse. Meant to replace the primary functions of format and parse of momentjs.

## NPM

```bash
$ npm install date-format-parse --save
```

## Usage

### Format

Format Date to String.

```js
import { format } from 'date-format-parse';

format(new Date(), 'YYYY-MM-DD HH:mm:ss.SSS');

// with locale, see locale config below
const obj = { ... }

format(new Date(), 'YYYY-MM-DD', { locale: obj })

```

### Parse

Parse String to Date

```js
import { parse } from 'date-format-parse';

parse('2019-12-10 14:11:12', 'YYYY-MM-DD HH:mm:ss'); // new Date(2019, 11, 10, 14, 11, 12)

// with backupDate, default is new Date()
parse('10:00', 'HH:mm', { backupDate: new Date(2019, 5, 6) }) // new Date(2019, 5, 6, 10)

// with locale, see locale config below
const obj = { ... }

parse('2019-12-10 14:11:12', 'YYYY-MM-DD HH:mm:ss', { locale: obj });

```

### Locale

```ts
interface Locale {
  months: string[];
  monthsShort: string[];
  weekdays: string[];
  weekdaysShort: string[];
  weekdaysMin: string[];
  meridiem?: (hours: number, minutes: number, isLowercase: boolean) => string;
  meridiemParse?: RegExp;
  isPM?: (input: string) => boolean;
  ordinal?: () => string;
}

const locale = {
  // MMMM
  months: [
    'January',
    'February',
    'March',
    'April',
    'May',
    'June',
    'July',
    'August',
    'September',
    'October',
    'November',
    'December',
  ],
  // MMM
  monthsShort: ['Jan', 'Feb', 'Mar', 'Apr', 'May', 'Jun', 'Jul', 'Aug', 'Sep', 'Oct', 'Nov', 'Dec'],
  // dddd
  weekdays: ['Sunday', 'Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday'],
  // ddd
  weekdaysShort: ['Sun', 'Mon', 'Tue', 'Wed', 'Thu', 'Fri', 'Sat'],
  // dd
  weekdaysMin: ['Su', 'Mo', 'Tu', 'We', 'Th', 'Fr', 'Sa'],
  // [A a] format the ampm. The following is the default value
  meridiem: (h: number, m: number, isLowercase: boolean) => {
    const word = h < 12 ? 'AM' : 'PM';
    return isLowercase ? word.toLocaleLowerCase() : word;
  };
  // [A a] used by parse to match the ampm. The following is the default value
  meridiemParse: /[ap]\.?m?\.?/i,
  // [A a] used by parse to determine if the matching string is pm. The following is the default value
  isPM: (input) => {
    return (input + '').toLowerCase().charAt(0) === 'p';
  }
};
```

## Tokens

| Uint                       | Token | output                                 |
| -------------------------- | ----- | -------------------------------------- |
| Year                       | YY    | 70 71 ... 10 11                        |
|                            | YYYY  | 1970 1971 ... 2010 2011                |
| Month                      | M     | 1 2 ... 11 12                          |
|                            | MM    | 01 02 ... 11 12                        |
|                            | MMM   | Jan Feb ... Nov Dec                    |
|                            | MMMM  | January February ... November December |
| Day of Month               | D     | 1 2 ... 30 31                          |
|                            | DD    | 01 02 ... 30 31                        |
| Day of Week                | d     | 0 1 ... 5 6                            |
|                            | dd    | Su Mo ... Fr Sa                        |
|                            | ddd   | Sun Mon ... Fri Sat                    |
|                            | dddd  | Sunday Monday ... Friday Saturday      |
| AM/PM                      | A     | AM PM                                  |
|                            | a     | am pm                                  |
| Hour                       | H     | 0 1 ... 22 23                          |
|                            | HH    | 00 01 ... 22 23                        |
|                            | h     | 1 2 ... 12                             |
|                            | hh    | 01 02 ... 12                           |
| Minute                     | m     | 0 1 ... 58 59                          |
|                            | mm    | 00 01 ... 58 59                        |
| Second                     | s     | 0 1 ... 58 59                          |
|                            | ss    | 00 01 ... 58 59                        |
| Fractional Second          | S     | 0 1 ... 8 9                            |
|                            | SS    | 00 01 ... 98 99                        |
|                            | SSS   | 000 001 ... 998 999                    |
| Time Zone                  | Z     | -07:00 -06:00 ... +06:00 +07:00        |
|                            | ZZ    | -0700 -0600 ... +0600 +0700            |
| Unix Timestamp             | X     | 1360013296                             |
| Unix Millisecond Timestamp | x     | 1360013296123                          |


CHANGELOG.md
## [0.2.4](https://github.com/mengxiong10/date-format-parse/compare/v0.2.3...v0.2.4) (2019-11-11)


### Bug Fixes

* negative year ([dad4e83](https://github.com/mengxiong10/date-format-parse/commit/dad4e833a885470e530f4d0165d0ed0de5b9d15e))



## [0.2.3](https://github.com/mengxiong10/date-format-parse/compare/v0.2.2...v0.2.3) (2019-11-10)


### Bug Fixes

* es build locale type ([969cab4](https://github.com/mengxiong10/date-format-parse/commit/969cab4056bc1b1b2cc6d7ddf3c6ca3d08a63033))



## [0.2.2](https://github.com/mengxiong10/date-format-parse/compare/v0.2.1...v0.2.2) (2019-11-10)


### Bug Fixes

* fix the parse types ([35c8ba3](https://github.com/mengxiong10/date-format-parse/commit/35c8ba31fe4b7aed28828236d7adda21d42ffe61))



## [0.2.1](https://github.com/mengxiong10/date-format-parse/compare/v0.2.0...v0.2.1) (2019-11-09)



# 0.2.0 (2019-11-09)

### Features

- add locales ([5467e51](https://github.com/mengxiong10/date-format-parse/commit/5467e5151f25c9bf055c0997ad8d346b6eb28c48))
- add the week token `w` `ww` ([388f09c](https://github.com/mengxiong10/date-format-parse/commit/388f09c2630805c4c107b09274d7148c5735079b))


format.js
var REGEX_FORMAT = /\[([^\]]+)]|YY(?:YY)?|M{1,4}|D{1,2}|d{1,4}|H{1,2}|h{1,2}|m{1,2}|s{1,2}|Z{1,2}|S{1,3}|a|A/g;
function pad(val, len) {
  if (len === void 0) {
    len = 2;
  }
  var output = Math.abs(val).toString();
  while (output.length < len) {
    output = '0' + output;
  }
  return output;
}
function formatTimezone(offset, delimeter) {
  if (delimeter === void 0) {
    delimeter = '';
  }
  var sign = offset > 0 ? '-' : '+';
  var absOffset = Math.abs(offset);
  var hours = Math.floor(absOffset / 60);
  var minutes = absOffset % 60;
  return sign + pad(hours, 2) + delimeter + pad(minutes, 2);
}
var defaultLocale = {
  months: ['January', 'February', 'March', 'April', 'May', 'June', 'July', 'August', 'September', 'October', 'November', 'December'],
  monthsShort: ['Jan', 'Feb', 'Mar', 'Apr', 'May', 'Jun', 'Jul', 'Aug', 'Sep', 'Oct', 'Nov', 'Dec'],
  weekdays: ['Sunday', 'Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday'],
  weekdaysShort: ['Sun', 'Mon', 'Tue', 'Wed', 'Thu', 'Fri', 'Sat'],
  weekdaysMin: ['Su', 'Mo', 'Tu', 'We', 'Th', 'Fr', 'Sa'],
};
var meridiem = function(hours, minutes, isLowercase) {
  var m = hours < 12 ? 'AM' : 'PM';
  return isLowercase ? m.toLocaleLowerCase() : m;
};
var formatters = {
  // Year: 00, 01, ..., 99
  YY: function(date) {
    return pad(date.getFullYear(), 4).substr(2);
  },
  // Year: 1900, 1901, ..., 2099
  YYYY: function(date) {
    return pad(date.getFullYear(), 4);
  },
  // Month: 1, 2, ..., 12
  M: function(date) {
    return date.getMonth() + 1;
  },
  // Month: 01, 02, ..., 12
  MM: function(date) {
    return pad(date.getMonth() + 1, 2);
  },
  MMM: function(date, locale) {
    return locale.monthsShort[date.getMonth()];
  },
  MMMM: function(date, locale) {
    return locale.months[date.getMonth()];
  },
  // Day of month: 1, 2, ..., 31
  D: function(date) {
    return date.getDate();
  },
  // Day of month: 01, 02, ..., 31
  DD: function(date) {
    return pad(date.getDate(), 2);
  },
  // Day of week: 0, 1, ..., 6
  d: function(date) {
    return date.getDay();
  },
  // Day of week: 'Su', 'Mo', ..., 'Sa'
  dd: function(date, locale) {
    return locale.weekdaysMin[date.getDay()];
  },
  // Day of week: 'Sun', 'Mon',..., 'Sat'
  ddd: function(date, locale) {
    return locale.weekdaysShort[date.getDay()];
  },
  // Day of week: 'Sunday', 'Monday', ...,'Saturday'
  dddd: function(date, locale) {
    return locale.weekdays[date.getDay()];
  },
  // Hour: 0, 1, ... 23
  H: function(date) {
    return date.getHours();
  },
  // Hour: 00, 01, ..., 23
  HH: function(date) {
    return pad(date.getHours(), 2);
  },
  // Hour: 1, 2, ..., 12
  h: function(date) {
    var hours = date.getHours();
    if (hours === 0) {
      return 12;
    } else if (hours > 12) {
      return hours % 12;
    } else {
      return hours;
    }
  },
  // Hour: 01, 02, ..., 12
  hh: function(date) {
    var hours = formatters['h'](date);
    return pad(hours, 2);
  },
  // Minute: 0, 1, ..., 59
  m: function(date) {
    return date.getMinutes();
  },
  // Minute: 00, 01, ..., 59
  mm: function(date) {
    return pad(date.getMinutes(), 2);
  },
  // Second: 0, 1, ..., 59
  s: function(date) {
    return date.getSeconds();
  },
  // Second: 00, 01, ..., 59
  ss: function(date) {
    return pad(date.getSeconds(), 2);
  },
  // 1/10 of second: 0, 1, ..., 9
  S: function(date) {
    return Math.floor(date.getMilliseconds() / 100);
  },
  // 1/100 of second: 00, 01, ..., 99
  SS: function(date) {
    return pad(Math.floor(date.getMilliseconds() / 10), 2);
  },
  // Millisecond: 000, 001, ..., 999
  SSS: function(date) {
    return pad(date.getMilliseconds(), 3);
  },
  // AM, PM
  A: function(date, locale) {
    var meridiemFunc = locale.meridiem || meridiem;
    return meridiemFunc(date.getHours(), date.getMinutes(), false);
  },
  // am, pm
  a: function(date, locale) {
    var meridiemFunc = locale.meridiem || meridiem;
    return meridiemFunc(date.getHours(), date.getMinutes(), true);
  },
  // Timezone: -01:00, +00:00, ... +12:00
  Z: function(date) {
    return formatTimezone(date.getTimezoneOffset(), ':');
  },
  // Timezone: -0100, +0000, ... +1200
  ZZ: function(date) {
    return formatTimezone(date.getTimezoneOffset());
  },
  // Seconds timestamp: 512969520
  X: function(date) {
    return Math.floor(date.getTime() / 1000);
  },
  // Milliseconds timestamp: 512969520900
  x: function(date) {
    return date.getTime();
  },
};
function format(val, str, options) {
  if (options === void 0) {
    options = {};
  }
  var formatStr = str ? String(str) : 'YYYY-MM-DDTHH:mm:ss.SSSZ';
  var date = val instanceof Date ? new Date(val.getTime()) : new Date(val);
  if (isNaN(date.getTime())) {
    return 'Invalid Date';
  }
  var locale = options.locale || defaultLocale;
  return formatStr.replace(REGEX_FORMAT, function(match, p1) {
    if (p1) {
      return p1;
    }
    if (typeof formatters[match] === 'function') {
      return '' + formatters[match](date, locale);
    }
    return match;
  });
}

const a = format(new Date(), 'YYYY-MM-DD');

console.log(a);
